src/H5.c

```diff
@@ -346,108 +356,137 @@ H5_term_library(void)
         H5_atclose_head = NULL;
     } /* end if */
 
+    /* clang-format off */
+
     /*
      * Terminate each interface. The termination functions return a positive
      * value if they do something that might affect some other interface in a
      * way that would necessitate some cleanup work in the other interface.
      */
-#define DOWN(F)                                                                                              \
-    (((n = H5##F##_term_package()) && (at + 8) < sizeof loop)                                                \
-         ? (HDsprintf(loop + at, "%s%s", (at ? "," : ""), #F), at += HDstrlen(loop + at), n)                 \
-         : ((n > 0 && (at + 5) < sizeof loop) ? (HDsprintf(loop + at, "..."), at += HDstrlen(loop + at), n)  \
-                                              : n))
 
-    do {
-        pending = 0;
-
-        /* Try to organize these so the "higher" level components get shut
-         * down before "lower" level components that they might rely on. -QAK
-         */
+#define TERMINATOR(module, wait) {      \
+      .func = H5##module##_term_package \
+    , .name = #module                   \
+    , .completed = false                \
+    , .await_prior = wait               \
+    }
 
+    /*
+     * Termination is ordered by the `terminator` table so the "higher" level
+     * packages are shut down before "lower" level packages that they
+     * rely on:
+     */
+    struct {
+        int (*func)(void);       /* function to terminate the module; returns 0
+                                  * on success, >0 if termination was not
+                                  * completed and we should try to terminate
+                                  * some dependent modules, first.
+                                  */
+        const char *name;        /* name of the module */
+        bool        completed;   /* true iff this terminator was already
+                                  * completed
+                                  */
+        const bool  await_prior; /* true iff all prior terminators in the list
```

Comment:
Internally we should be using hbool_t for consistency. I'd like to strip hbool_t from the library, but until then we should continue to use it.